[DeviceMesh][ez] Extract the pg creation as a util function

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #163367
* __->__ #163930
* #163928
* #163288
* #163212

This is just to extract common logic into a util function because we will use it many times for the following stack of Device Mesh refactoring.

cc @H-Huang @awgu @wanchaol @fegin @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci

fduwjj committed:
[DeviceMesh][ez] Extract the pg creation as a util function

[ghstack-poisoned]

fegin commented on torch/distributed/device_mesh.py:
```diff
@@ -856,6 +860,54 @@ def get_all_groups(self) -> list[ProcessGroup]:
+        ) -> "DeviceMesh":
+            """
+            Helper method to create a DeviceMesh from tensor `pg_ranks_by_dim`. This is due to
+            the constraint of ProcessGroup API that all ranks have to call the PG creation API.
```
```suggestion
            the constraint of ProcessGroup API that all ranks have to call the PG creation API even if the rank is not in that PG.
```

fduwjj committed:
Update on "[DeviceMesh][ez] Extract the pg creation as a util function"


This is just to extract common logic into a util function because we will use it many times for the following stack of Device Mesh refactoring.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

lw commented on torch/distributed/device_mesh.py:
```diff
@@ -856,6 +860,55 @@ def get_all_groups(self) -> list[ProcessGroup]:
+            """
+            res_mesh = None
+            for mesh_nd in pg_ranks_by_dim:
+                mesh = DeviceMesh(
+                    device_type,
+                    mesh_nd,
+                    mesh_dim_names=mesh_dim_names,
+                    backend_override=backend_override,
+                    _init_backend=_init_backend,
+                )
+                if cur_rank in mesh_nd:
+                    res_mesh = mesh
```
Thanks for splitting this out as I had asked! I didn't get to review this before it was landed, but there's still something I don't understand. I get that we need to call the "PG creation API" multiple times on each rank, even for the ranks that don't participate, but I don't get why we need to call the _DeviceMesh_ constructor multiple times!

Could we instead call the PG creation API directly and just invoke the DeviceMesh constructor once with the right PG?
